### ISSUE TITLE
Android: Sharing Flutter Engine between foreground and background isolates

### ISSUE THREAD

**Issue by ebajrami** [flutter/flutter#60198]
## Use case
We want to start the app based on a background trigger on Android.

We have a long running _foreground service_ which is started based on the Bluetooth connection state. Since the service can be started when the app is not running, the service needs to get the information from the Flutter (e.g. access token and refresh token).

We want to start an isolate (from the foreground service) which can be later associated with the main app view in order to prevent multiple instance creations (e.g. authentication, access to the file system, external libs, etc). Currently we are able to start two different isolates (foreground and background) but there are synchronisation issues between those two. Also, when the app is then put to background we want to keep the isolate running in a foreground service.

For example:
- Bluetooth broadcast is received
- We start the background service (Android native) and instantiate the Dart isolate
- User starts the app
- UI isolate is started

We tried to kill the background isolate as soon the app is started but this causes other issues.

There is also this Meta issue that seems related: https://github.com/flutter/flutter/issues/32164

## Proposal
If we could use the isolate which is created in the background as the main Flutter instance as soon as the app is started would solve our issue.

**Comment by zoeyfan** [comment]
@chinmaygarde This is the issue filed by the customer mentioned in the customer question doc. Just want to bring it to your attention.

**Comment by zoeyfan** [comment]
Bumping priority to P2 as this comes from customer Vroom.

**Comment by chinmaygarde** [comment]
The Mobile platforms team is overloaded ATM and probably cannot address this issue this quarter (@xster for confirmation). It may be the case that the Engine team needs to figure out an owner for this.

**Comment by zoeyfan** [comment]
Cc @csells

**Comment by xster** [comment]
Yes, the thing to solve here is the general background execution meta issue https://github.com/flutter/flutter/issues/32164. Productionizing a general solution will likely be a quarter out if it's prioritized. In the meantime, we might be able to give you some suggestions for work-arounds in your own project (it's a general Android dependency injection solution you need rather than a flutter "background" solution).

**Comment by afrueh** [comment]
We're currently investigating to separate the boot-up in Android in 2 steps if the app is launched from the background:

1. Have a custom (Generated)PluginRegistrant where dependencies are loaded that are required by the background service and not load any dependencies that require some sort of a view

2. As soon as we go to foreground we're trying to share the instance of the Flutter VM that was created in the background with the main application and load all remaining Plugins / Dependencies

Is this a potential work-around for the issue?
Do you see any problems with an approach like this?

**Comment by xster** [comment]
> Have a custom (Generated)PluginRegistrant where dependencies are loaded

I'm assuming you're just talking about plugins right? Sure that sounds reasonable.

> we're trying to share the instance of the Flutter VM

Concretely I'm assuming you're referring to the FlutterEngine?

> load all remaining Plugins / Dependencies

You might be already taking care of this. Note [this line](https://cs.opensource.google/flutter/engine/+/master:shell/platform/android/io/flutter/embedding/android/FlutterActivity.java;l=875) loads all plugins. If you don't exclude the ones already registered, you might end up creating and registering 2 instances of the same plugin. 

Generally, that's the right approach.

**Comment by gaaclarke** [comment]
Forgive me I don't know Android super well, but why can't they just keep one isolate without a view that runs as a Service, then when the app is launched they start a new isolate that just uses IPC to communicate with the Service?  Then there would be no need for swapping or conditional loading of plugins.

**Comment by chinmaygarde** [comment]
Triage: Assigning to @zanderso for followup.

**Comment by zanderso** [comment]
I don't think we'll be able to prioritize the productionization of general background execution at this time, so approaches to solving this specific instance are preferable.

**Comment by Hixie** [comment]
I did something similar to this in my Rainbow Monkey app using android_alarm_manager:
   https://github.com/jocosocial/rainbowmonkey/blob/master/lib/src/logic/background_polling.dart
I'm not sure how much that will help but maybe looking at that code and the code in android_alarm_manager itself might help?

**Comment by zanderso** [comment]
I'm worried that my last comment might not have been as clear as it could have been. It sounds like we can't prioritize solving https://github.com/flutter/flutter/issues/32164 in its full generality in the near-term, but we can still help come up with a solution for this issue. It looks like the next step is to get the questions from @xster and @gaaclarke answers from @afrueh or @ebajrami. Sorry for any confustion! /cc @csells

**Comment by xster** [comment]
Yes, if you still have questions in your specific Android app in getting the workaround we described to work, we can help you. I believe the workaround (doing standard Android dependency injection, probably via the Application) should be feasible.

**Comment by afrueh** [comment]
Hello,
sorry for the late reply - we're still working on getting everything to work, right now it looks like it is going on well 

We're still waiting on a 3rd party to adapt their plugin to not require a 'view' anymore. After that we can move forward and start testing the background code in a bigger scope.

@ebajrami could you please sync with the Vincent and post here if we require more support or if everything is working?

Thanks a lot for your help.

**Comment by xster** [comment]
Cool. Good to hear. Let us know if you need anything. 

> waiting on a 3rd party to adapt their plugin to not require a 'view' anymore

Yes that's an unfortunate state of affairs. Thus we needed to move to newer Android embedding APIs fast last year to not let more privilege leaks accumulate in the ecosystem :S

**Comment by chinmaygarde** [comment]
Reducing the priority of this issue per the last comment.

**Comment by ebajrami** [comment]
> Hello,
> sorry for the late reply - we're still working on getting everything to work, right now it looks like it is going on well
> 
> We're still waiting on a 3rd party to adapt their plugin to not require a 'view' anymore. After that we can move forward and start testing the background code in a bigger scope.
> 
> @ebajrami could you please sync with the Vincent and post here if we require more support or if everything is working?
> 
> Thanks a lot for your help.

As you mentioned, we are waiting for the 3rd party plugins to adopt their plugins so we can proceed with testing of our changes. We will update the ticket once we have some initial results.

**Comment by MartinHlavna** [comment]
@ebajrami Hello, did you please got that workaround working? 

@xster , @afrueh Could you be please more elaborate on workaround details? We are trying to achieve something similar. We have code triggered by   `WorkManager` and also have synchronization problems between isolates. We have failed to solve them by `SendPort`/`ReceivePort` because of race conditions. 

Now we are trying to share `FlutterEngine`. We managed to get `FlutterEngineCache` working but when executing `DartCallback` it fails with error "Attempted to run a DartExecutor that is already running."

Could you give as a few pointers where to look next?

**Comment by xster** [comment]
> We have failed to solve them by SendPort/ReceivePort because of race conditions.

What race conditions are you encountering specifically? The general approach seems feasible. 

> We managed to get FlutterEngineCache working but when executing DartCallback it fails with error "Attempted to run a DartExecutor that is already running."

Presumably you already got that FlutterEngine instance running the first time when you were in the background. You wouldn't run it a second time but instead, would attach it to a FlutterActivity or FlutterFragment by using the appropriate intent builder (assuming you already called runApp in Dart when you were in the background).

**Comment by kheuss** [comment]
I want to set up a mobile application with flutter which also runs in the background. this application allows you to scan bluetooth devices and listen to events to launch a notification and / or start a ringtone. I managed to do all this and it works very well with the flutter_blue plugin. But my problem is that the application has to keep running in the background.

I came here to seek help.

The app does exactly what this app does https://play.google.com/store/apps/details?id=com.antilost.app3&hl=fr&gl=US

**Comment by JulioCRFilho** [comment]
Did you guys get any update on this?

I'm doing something similar: I have 2 custom keyboard extensions (Android and iOS), and they run separately from the Flutter part. Then as the keyboard is initiated, we run the Flutter part with FlutterEngine.

The problem is: When the Flutter part runs, it has its own plugins and their platform part that needs to run and initialise the native plugins, register the flutterChannel for these plugins and communicate with the Flutter main app, that will return the value from these plugins to the Flutter side, and then to the keyboard extension. But any plugin we try to do this, throws an Exception of method not implemented.

I'm not quite sure this is the same issue, but when we run it by the Flutter side it works perfectly, which I assume is because all the plugins use the same UI Isolate then. But the extension goal is to the user don't have to leave whatever app his on to be able to execute our app's methods.

**Comment by MartinHlavna** [comment]
@JulioCRFilho 

we have kinda worked around (to be honest it is more like ugly hack) by having one FlutterEngine (managed by Application class) that holds any shared data and acts as synchonization semaphor. All comunitcation is done using IsolateNameServer and SendPort/ReceivePort combo. 

Since FlutterActivity and BackgroundWorker both spawns their respective engines it was only way we could ensure no conflicts between them.

**Comment by JulioCRFilho** [comment]
> @JulioCRFilho
> 
> we have kinda worked around (to be honest it is more like ugly hack) by having one FlutterEngine (managed by Application class) that holds any shared data and acts as synchonization semaphor. All comunitcation is done using IsolateNameServer and SendPort/ReceivePort combo.
> 
> Since FlutterActivity and BackgroundWorker both spawns their respective engines it was only way we could ensure no conflicts between them.

Can you elaborate about this implementation? I'm kinda new to this stuff, so I'm not sure I got it.
My applications run directly from the platform part invoking the Flutter part, do you think this approach is available
for this case?

**Comment by JulioCRFilho** [comment]
@MartinHlavna sorry for bother you with this, but I tried the Geofencing implementation to communicate with a Plugin in background using the PluginUtilities.getCallbackHandle, but still the same method not implemented exception.
Is there any kind of guide for the work around you managed to do?
